test/common/event/scaled_range_timer_manager_test.cc

```diff
@@ -0,0 +1,450 @@
+      std::make_unique<std::vector<MonotonicTime>>()};
+  RangeTimerPtr timer;
+};
+
+TEST_F(ScaledRangeTimerManagerTest, CreateAndDestroy) {
+  ScaledRangeTimerManager manager(dispatcher_, 1.0);
+}
+
+TEST_F(ScaledRangeTimerManagerTest, CreateAndDestroyTimer) {
+  ScaledRangeTimerManager manager(dispatcher_, 1.0);
+
+  {
+    StrictMock<MockFunction<TimerCb>> callback;
+    auto timer = manager.createTimer(callback.AsStdFunction());
+  }
+}
+
+TEST_F(ScaledRangeTimerManagerTest, CreateSingleScaledTimer) {
+  ScaledRangeTimerManager manager(dispatcher_, 1.0);
+
+  StrictMock<MockFunction<TimerCb>> callback;
+  auto timer = manager.createTimer(callback.AsStdFunction());
+  EXPECT_CALL(callback, Call());
+
+  timer->enableTimer(std::chrono::seconds(5), std::chrono::seconds(10));
+  EXPECT_TRUE(timer->enabled());
+
+  simTime().advanceTimeAsync(std::chrono::seconds(5));
+  dispatcher_.run(Dispatcher::RunType::Block);
+  EXPECT_TRUE(timer->enabled());
+
+  simTime().advanceTimeAsync(std::chrono::seconds(5));
+  dispatcher_.run(Dispatcher::RunType::Block);
+  EXPECT_FALSE(timer->enabled());
+}
+
+TEST_F(ScaledRangeTimerManagerTest, EnableAndDisableTimer) {
+  ScaledRangeTimerManager manager(dispatcher_, 1.0);
+
+  StrictMock<MockFunction<TimerCb>> callback;
+  auto timer = manager.createTimer(callback.AsStdFunction());
+
+  timer->enableTimer(std::chrono::seconds(5), std::chrono::seconds(30));
+  EXPECT_TRUE(timer->enabled());
+
+  timer->disableTimer();
+  EXPECT_FALSE(timer->enabled());
+
+  // Provide some additional guarantee of safety by running the dispatcher for a little bit. This
+  // should be a no-op, and if not (because a timer was fired), that's a problem that will be caught
+  // by the strict mock callback.
+  simTime().advanceTimeAsync(std::chrono::seconds(10));
+  dispatcher_.run(Dispatcher::RunType::Block);
+}
+
+TEST_F(ScaledRangeTimerManagerTest, DisableWhileDisabled) {
+  ScaledRangeTimerManager manager(dispatcher_, 1.0);
+
+  StrictMock<MockFunction<TimerCb>> callback;
+  auto timer = manager.createTimer(callback.AsStdFunction());
+
+  EXPECT_FALSE(timer->enabled());
+  timer->disableTimer();
+
+  EXPECT_FALSE(timer->enabled());
+}
+
+TEST_F(ScaledRangeTimerManagerTest, DisableWhilePending) {
+  ScaledRangeTimerManager manager(dispatcher_, 1.0);
+
+  StrictMock<MockFunction<TimerCb>> callback;
+  auto timer = manager.createTimer(callback.AsStdFunction());
+  timer->enableTimer(std::chrono::seconds(10), std::chrono::seconds(100));
+  EXPECT_TRUE(timer->enabled());
+
+  timer->disableTimer();
+  EXPECT_FALSE(timer->enabled());
+}
+
+TEST_F(ScaledRangeTimerManagerTest, DisableWhileActive) {
+  ScaledRangeTimerManager manager(dispatcher_, 1.0);
+
+  StrictMock<MockFunction<TimerCb>> callback;
+  auto timer = manager.createTimer(callback.AsStdFunction());
+
+  timer->enableTimer(std::chrono::seconds(5), std::chrono::seconds(100));
+
+  simTime().advanceTimeAsync(std::chrono::seconds(5));
```

Comment:
Worth adding some assertions about the queue state after the advance and run?

Reply:
It's not part of the public interface, so I'm hesitant to test it directly.